core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp

```diff
@@ -0,0 +1,724 @@
+//@HEADER
+*/
+
+#ifndef KOKKOS_OPENACC_PARALLEL_FOR_MDRANGE_HPP
+#define KOKKOS_OPENACC_PARALLEL_FOR_MDRANGE_HPP
+
+#include <OpenACC/Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_FunctorAdapter.hpp>
+#include <OpenACC/Kokkos_OpenACC_MDRangePolicy.hpp>
+#include <Kokkos_Parallel.hpp>
+
+namespace Kokkos::Experimental::Impl {
+
+struct OpenACCCollapse {};
+struct OpenACCTile {};
+using OpenACCIterateLeft  = std::integral_constant<Iterate, Iterate::Left>;
+using OpenACCIterateRight = std::integral_constant<Iterate, Iterate::Right>;
+template <int N>
+using OpenACCMDRangeBegin = decltype(MDRangePolicy<OpenACC, Rank<N>>::m_lower);
+template <int N>
+using OpenACCMDRangeEnd = decltype(MDRangePolicy<OpenACC, Rank<N>>::m_upper);
+template <int N>
+using OpenACCMDRangeTile = decltype(MDRangePolicy<OpenACC, Rank<N>>::m_tile);
+
+template <class Functor>
+void OpenACCParallelForMDRangePolicy(OpenACCCollapse, OpenACCIterateLeft,
+                                     Functor const& functor,
+                                     OpenACCMDRangeBegin<2> const& begin,
+                                     OpenACCMDRangeEnd<2> const& end,
+                                     int async_arg) {
+  int begin1 = begin[1];
+  int end1   = end[1];
+  int begin0 = begin[0];
+  int end0   = end[0];
+// clang-format off
+#pragma acc parallel loop gang vector collapse(2) copyin(functor) async(async_arg)
+  // clang-format on
+  for (auto i1 = begin1; i1 < end1; ++i1) {
+    for (auto i0 = begin0; i0 < end0; ++i0) {
+      functor(i0, i1);
+    }
+  }
+}
+
+template <class Functor>
+void OpenACCParallelForMDRangePolicy(OpenACCCollapse, OpenACCIterateRight,
+                                     Functor const& functor,
+                                     OpenACCMDRangeBegin<2> const& begin,
+                                     OpenACCMDRangeEnd<2> const& end,
+                                     int async_arg) {
+  int begin0 = begin[0];
+  int end0   = end[0];
+  int begin1 = begin[1];
+  int end1   = end[1];
+// clang-format off
+#pragma acc parallel loop gang vector collapse(2) copyin(functor) async(async_arg)
+  // clang-format on
+  for (auto i0 = begin0; i0 < end0; ++i0) {
+    for (auto i1 = begin1; i1 < end1; ++i1) {
+      functor(i0, i1);
+    }
+  }
+}
+
+template <class Functor>
+void OpenACCParallelForMDRangePolicy(OpenACCTile, OpenACCIterateLeft,
+                                     Functor const& functor,
+                                     OpenACCMDRangeBegin<2> const& begin,
+                                     OpenACCMDRangeEnd<2> const& end,
+                                     OpenACCMDRangeTile<2> const& tile,
+                                     int async_arg) {
+  int tile0  = tile[0];
+  int tile1  = tile[1];
+  int begin1 = begin[1];
+  int end1   = end[1];
+  int begin0 = begin[0];
+  int end0   = end[0];
+// clang-format off
+#pragma acc parallel loop gang vector tile(tile0,tile1) copyin(functor) async(async_arg)
```

Comment:
OpenACC says that the tile size should be a constant positive integer expression, but `tile0` and `tile1` are not constant.

Reply:
runtime or compile-time constant?

Reply:
compile-time constant.